plugins/inputs/elasticsearch_query/elasticsearch_query.go

```diff
@@ -0,0 +1,277 @@
+  ## Set to true to not ignore documents when the tag(s) above are missing
+  include_missing_tag = true
+  ## String value of the tag when the tag does not exist
+  ## Used when include_missing_tag is true
+  missing_tag_value = "null"
+`
+
+type ElasticsearchQuery struct {
+	URLs                []string `toml:"urls"`
+	Username            string
+	Password            string
+	EnableSniffer       bool
+	Tracelog            bool
+	Timeout             internal.Duration
+	HealthCheckInterval internal.Duration
+	Aggregations        []Aggregation `toml:"aggregation"`
+	tls.ClientConfig
+	httpclient *http.Client
+	ESClient   *elastic.Client
+	acc        telegraf.Accumulator
+}
+
+type Aggregation struct {
+	Index             string
+	MeasurementName   string
+	FilterQuery       string
+	QueryPeriod       internal.Duration
+	MetricFields      []string `toml:"metric_fields"`
+	DateField         string
+	Tags              []string `toml:"tags"`
+	IncludeMissingTag bool
+	MissingTagValue   string
+	MetricFunction    string
+}
+
+type aggKey struct {
+	measurement string
+	name        string
+	function    string
+	field       string
+}
+
+type aggregationQueryData struct {
+	aggKey
+	isParent    bool
+	aggregation elastic.Aggregation
+}
+
+func (e *ElasticsearchQuery) SampleConfig() string {
+	return sampleConfig
+}
+
+func (e *ElasticsearchQuery) Description() string {
+	return description
+}
+
+func (e *ElasticsearchQuery) init() error {
+	if e.URLs == nil {
+		return fmt.Errorf("Elasticsearch urls is not defined")
+	}
+
+	err := e.connectToES()
+	return err
+}
+
+func (e *ElasticsearchQuery) connectToES() error {
+
+	var clientOptions []elastic.ClientOptionFunc
+
+	if e.httpclient == nil {
+		httpclient, err := e.createHttpClient()
+		if err != nil {
+			return err
+		}
+		e.httpclient = httpclient
+	}
+
+	clientOptions = append(clientOptions,
+		elastic.SetHttpClient(e.httpclient),
+		elastic.SetSniff(e.EnableSniffer),
+		elastic.SetURL(e.URLs...),
+		elastic.SetHealthcheckInterval(e.HealthCheckInterval.Duration),
+	)
+
+	if e.Username != "" && e.Password != "" {
+		clientOptions = append(clientOptions,
+			elastic.SetBasicAuth(e.Username, e.Password),
+		)
+	}
+
+	if e.HealthCheckInterval.Duration == 0 {
+		clientOptions = append(clientOptions,
+			elastic.SetHealthcheck(false),
+		)
+	}
+
+	if e.Tracelog {
+		clientOptions = append(clientOptions,
+			elastic.SetTraceLog(log.New(os.Stdout, "", log.LstdFlags)),
+		)
+	}
+
+	client, err := elastic.NewClient(clientOptions...)
+	if err != nil {
+		return err
+	}
+
+	// check for ES version on first node
+	esVersion, err := client.ElasticsearchVersion(e.URLs[0])
+
+	if err != nil {
+		return fmt.Errorf("Elasticsearch query version check failed: %s", err)
+	}
+
+	// quit if ES version is not supported
+	i, err := strconv.Atoi(strings.Split(esVersion, ".")[0])
+	if err != nil || i < 5 {
+		return fmt.Errorf("Elasticsearch query: ES version not supported: %s", esVersion)
```

Comment:
Errors have to start lower-case. Please also change the other instances. Furthermore it would be nice here to specify the supported version. Something like "ES versions < 5.x are not supported".